fe1-web/network/NetworkConnection.ts

```diff
@@ -0,0 +1,155 @@
+import { w3cwebsocket as W3CWebSocket } from 'websocket';
+import {
+  JsonRpcRequest, JsonRpcResponse, ProtocolError, UNDEFINED_ID,
+} from 'model/network';
+import { NetworkError } from './NetworkError';
+
+type ResponseHandler = (message: JsonRpcResponse) => void;
+
+const DEFAULT_RESPONSE_HANDLER: ResponseHandler = (message: JsonRpcResponse) => {
+  if (message.error === undefined) {
+    // default handler does nothing which is useful for some messages
+  } else {
+    console.error(
+      'A negative network error was received:\n'
+      + `\t- error code : ${message.error.code}`
+      + `\t- description : ${message.error.description}`,
+    );
+  }
+};
+
+const WEBSOCKET_READYSTATE_INTERVAL_MS = 10;
+const WEBSOCKET_READYSTATE_MAX_ATTEMPTS = 100;
+
+const WEBSOCKET_MESSAGE_TIMEOUT_MS = 10000; // 10 seconds max round-trip time
+
+interface PendingResponse {
+  promise: Promise<JsonRpcResponse>,
+  resolvePromise: (value: JsonRpcResponse) => void,
+  rejectPromise: () => void,
+  timeoutId: NodeJS.Timeout,
+}
+
+export class NetworkConnection {
+  private readonly ws: any;
+
+  private payloadSentCount: number = 0;
+
+  private payloadPending: Map<number, PendingResponse> = new Map();
+
+  private onMessageHandler: ResponseHandler = DEFAULT_RESPONSE_HANDLER;
+
+  public readonly address: string; // FIXME get the address from ws (log ws when app is running)
+
+  constructor(address: string) {
+    const ws = new W3CWebSocket(address);
+
+    ws.onopen = () => { console.info(`Initiating web socket : ${address}`); };
+    ws.onmessage = (message: any) => {
+      console.log(`Received a new message from '${address}' : `, message.data);
+
+      let response: JsonRpcResponse;
+      try {
+        response = JsonRpcResponse.fromJson(message.data);
+      } catch (e) {
+        throw new ProtocolError('Answer is not valid Json');
+      }
+
+      if (response.id !== UNDEFINED_ID) {
+        const pendingResponse = this.payloadPending.get(response.id);
+        if (pendingResponse === undefined) {
+          throw new NetworkError(`Received a response which id = ${response.id} does not match any pending requests`);
+        }
+
+        this.payloadPending.delete(response.id);
+
+        pendingResponse.resolvePromise(response);
```

Comment:
resolve or reject, based on whether the response is a result or a failure.
for convenience, it might make sense to have a getter/helper "isError" on the JsonRpcResponse, to keep the logic very clean here